File: src/main/java/org/proshin/blog/socialnetwork/instagram/oauth/RedirectEndpoint.java
package org.proshin.blog.socialnetwork.instagram.oauth;

import java.util.HashMap;
import java.util.Map;
import lombok.AllArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.proshin.blog.configuration.ConfigParameters;
import org.proshin.blog.dao.AccessTokens;
import static org.proshin.blog.model.AccessToken.Provider.INSTAGRAM;
import org.proshin.blog.socialnetwork.instagram.UsersEndpoint;
import org.proshin.blog.socialnetwork.instagram.model.User;
import org.proshin.blog.socialnetwork.instagram.oauth.response.AccessToken;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.client.RestTemplate;

@Slf4j
@Controller
@AllArgsConstructor
public class RedirectEndpoint {
    private static final String ACCESS_TOKEN_ENDPOINT = "https://api.instagram.com/oauth/access_token";

    private final ConfigParameters configParameters;
    private final AccessTokens accessTokens;
    private final RestTemplate restTemplate;

    @RequestMapping(value = "/oauth/instagram/redirect")
    public ResponseEntity<?> processCodeForInstagram(@RequestParam(name = "code", required = true) String code) {
        Map<String, String> request = new HashMap<>();
        request.put("client_id", configParameters.getInstagram().getClientId());
        request.put("client_secret", configParameters.getInstagram().getClientSecret());
        request.put("grant_type", "authorization_code");
        request.put("redirect_uri", configParameters.getInstagram().getRedirectUri());
        request.put("code", code);
        ResponseEntity<String> response = restTemplate.postForEntity(ACCESS_TOKEN_ENDPOINT, request, String.class);
        if (response.getStatusCode() == HttpStatus.OK) {
            org.proshin.blog.model.AccessToken accessToken =
                    accessTokens.create(INSTAGRAM,
                            new AccessToken(response.getBody()).getAccessToken());
            log.info("Access token has been successfully received: {}", accessToken);

            User user =
                    new UsersEndpoint(accessToken.getToken(), restTemplate)
                            .getSelf();
            log.info("User has been received: {}", user);
            return ResponseEntity.ok(user);
        }

        throw new Error(response.getBody());
    }
}

File: src/main/kotlin/org/proshin/blog/oauth/instagram/RedirectEndpoint.kt
package org.proshin.blog.oauth.instagram

import com.github.kittinunf.fuel.Fuel
import com.github.kittinunf.fuel.core.FuelError
import com.github.kittinunf.fuel.core.Handler
import com.github.kittinunf.fuel.core.Request
import com.github.kittinunf.fuel.core.Response
import mu.KotlinLogging
import org.proshin.blog.configuration.ConfigParameters
import org.proshin.blog.dao.AccessTokens
import org.proshin.blog.model.AccessToken.Provider.INSTAGRAM
import org.proshin.blog.oauth.instagram.response.AccessToken
import org.proshin.blog.oauth.instagram.response.Error
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.ResponseEntity
import org.springframework.stereotype.Controller
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam

@Controller
open class RedirectEndpoint {

    private val ACCESS_TOKEN_ENDPOINT = "https://api.instagram.com/oauth/access_token"
    private val logger = KotlinLogging.logger {}

    @Autowired
    lateinit var configParameters: ConfigParameters;
    @Autowired
    lateinit var accessTokens: AccessTokens

    @RequestMapping(value = "/oauth/instagram/redirect")
    fun processCodeForInstagram(@RequestParam(name = "code", required = true) code: String): ResponseEntity<String> {
        Fuel.post(ACCESS_TOKEN_ENDPOINT,
                listOf(
                        "client_id" to configParameters.instagram.clientId,
                        "client_secret" to configParameters.instagram.clientSecret,
                        "grant_type" to "authorization_code",
                        "redirect_uri" to configParameters.instagram.redirectUri,
                        "code" to code))
                .responseString(object : Handler<String> {
                    override fun failure(request: Request, response: Response, error: FuelError) {
                        val error = Error(String(response.data))
                        logger.error("Error has happened $error")
                    }

                    override fun success(request: Request, response: Response, value: String) {
                        val accessToken = accessTokens.create(INSTAGRAM, AccessToken(value).accessToken)
                        logger.info("Access token has been successfully received: $accessToken")
                    }
                })
        return ResponseEntity.ok().build()
    }
}

